Sources/Sentry/SentryReachability.m

```diff
@@ -103,60 +124,82 @@ @implementation SentryReachability
-    });
+        _sentry_reachability_ref = SCNetworkReachabilityCreateWithName(NULL, "sentry.io");
+        if (!_sentry_reachability_ref) { // Can be null if a bad hostname was specified
+            return;
+        }
 
-    _sentry_reachability_ref = SCNetworkReachabilityCreateWithName(NULL, "sentry.io");
-    if (!_sentry_reachability_ref) { // Can be null if a bad hostname was specified
-        return;
+        SENTRY_LOG_DEBUG(@"registering callback for reachability ref %@", _sentry_reachability_ref);
+        if (self.setReachabilityCallback) {
+            SCNetworkReachabilitySetCallback(
+                _sentry_reachability_ref, SentryConnectivityCallback, NULL);
+        }
+        SCNetworkReachabilitySetDispatchQueue(_sentry_reachability_ref, sentry_reachability_queue);
     }
-
-    SENTRY_LOG_DEBUG(@"registering callback for reachability ref %@", _sentry_reachability_ref);
-    SCNetworkReachabilitySetCallback(_sentry_reachability_ref, SentryConnectivityCallback, NULL);
-    SCNetworkReachabilitySetDispatchQueue(_sentry_reachability_ref, reachabilityQueue);
 }
 
 - (void)removeObserver:(id<SentryReachabilityObserver>)observer
 {
-    [sentry_reachability_change_blocks removeObjectForKey:[observer description]];
-    if (sentry_reachability_change_blocks.allValues.count > 0) {
+    SENTRY_LOG_DEBUG(@"Removing observer: %@", observer);
+    @synchronized(sentry_reachability_observers) {
+        SENTRY_LOG_DEBUG(@"Entered synchronized region of removeObserver");
+        [sentry_reachability_observers removeObject:observer];
+
+        [self unsetCallbackIfNeeded];
+    }
+}
+
+- (void)removeAllObservers
+{
+    @synchronized(sentry_reachability_observers) {
+        [sentry_reachability_observers removeAllObjects];
+        [self unsetCallbackIfNeeded];
+    }
+}
+
+- (void)unsetCallbackIfNeeded
+{
+    if (sentry_reachability_observers.count > 0) {
+        SENTRY_LOG_DEBUG(
+            @"Other observers still registered, will not unset reachability callback.");
         return;
     }
 
     sentry_current_reachability_state = kSCNetworkReachabilityFlagsUninitialized;
 
-    if (_sentry_reachability_ref == nil) {
-        SENTRY_LOG_WARN(@"No reachability ref to unregister.");
-        return;
+    if (_sentry_reachability_ref != nil) {
+        SENTRY_LOG_DEBUG(@"removing callback for reachability ref %@", _sentry_reachability_ref);
+        SCNetworkReachabilitySetCallback(_sentry_reachability_ref, NULL, NULL);
+        SCNetworkReachabilitySetDispatchQueue(_sentry_reachability_ref, NULL);
+        _sentry_reachability_ref = nil;
     }
 
-    SENTRY_LOG_DEBUG(@"removing callback for reachability ref %@", _sentry_reachability_ref);
-    SCNetworkReachabilitySetCallback(_sentry_reachability_ref, NULL, NULL);
-    SCNetworkReachabilitySetDispatchQueue(_sentry_reachability_ref, NULL);
+    SENTRY_LOG_DEBUG(@"Cleaning up reachability queue.");
+    sentry_reachability_queue = nil;
```

Comment:
I changed this so that even if the reachability ref is somehow nil, we still release the queue memory. Looking again, it looks like because of threading, this method could be called multiple times, so the first time it'd be nonnil and the next one would be nil. In any case this should be fine.